tests/kernel/ulp_openssl.pm

```diff
@@ -0,0 +1,178 @@
+        # If no targets found, end the test normally.
+        return if not defined($tinfo);
+    }
+
+    my $run_id = $tinfo->{run_id};
+    my $target_ver = $tinfo->{target_versions}[$run_id];
+    my $packname = $tinfo->{packname};
+    my $total_runs = scalar(@{$tinfo->{target_versions}});
+
+    record_info('Test Iter', "Iteration " . ($run_id + 1) . "/$total_runs: Testing with openssl-3-$target_ver");
+    # Ensure livepatch is not installed from a previous run
+    if (script_run("rpm -q $packname") == 0) {
+        uninstall_package($packname);
+    }
+    record_info('Downgrade', "Installing old version: openssl-3-$target_ver");
+    # Force install the old package. Exit codes 106/107 indicate updates are available, which is expected.
+    enter_cmd("zypper in --oldpackage -f openssl-3=$target_ver");
+    wait_serial(qr/Choose from above solutions.*/, 100);
+    type_string '1';
+    type_string("\n");
+    wait_serial(qr/Continue.*/, 100);
+    type_string 'y';
+    type_string("\n");
+
+    # Set LD_PRELOAD for libpulp, required for userspace livepatching.
+    assert_script_run('export LD_PRELOAD=/usr/lib64/libpulp.so.0');
+
+    # Generate temporary SSL certificates for the server.
+    assert_script_run("openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 1 -nodes -subj '/CN=localhost' -passout pass:password");
+
+    # Start `openssl s_server` in the background. It's a long-running process that links with libssl/libcrypto.
+    # We use enter_cmd and wait_serial to reliably get its PID.
+    enter_cmd("openssl s_server -cert cert.pem -key key.pem -pass pass:password -accept 44330 -www & echo PID=\$!.");
+    my $server_pid_line = wait_serial(qr/PID=\d+\./);
+    die 'Failed to parse server PID' unless $server_pid_line =~ m/PID=(\d+)\./;
+    my $server_pid = $1;
+    record_info('Workload', "Started openssl s_server with PID $server_pid");
+
+    # Verify the server is running and unpatched by checking its memory maps.
+    # It should point to the standard /usr/lib64 locations.
+    my $maps_before = script_output("cat /proc/$server_pid/maps | grep 'libcrypto\\|libssl'");
+    if ($maps_before =~ m!/var/(lib/)?livepatch!) {
+        die "Process is already patched before livepatch installation! Maps:\n$maps_before";
+    }
+
+    record_info('Install LP', "Installing livepatch package: $packname");
+    install_package($packname);
+
+    # Get the list of livepatch files (.so) installed by the package.
+    my $patch_files = script_output("rpm -ql $packname | grep '\\.so\$'");
+
+    record_info('Trigger', "Triggering livepatches...");
+    # Iterate over all patches and try to trigger them on the process.
+    # `ulp trigger` will apply the patch that matches the process's loaded libraries.
+    for my $patch (split(/\n/, $patch_files)) {
+        # We use a short timeout and ignore failures, as some patches might not be for our process.
+        script_run("time ulp trigger -r 10 --timeout 20 '$patch'");
```

Comment:
Livepatches should be already applied by `install_package($packname);`. Running `ulp trigger` is not needed and may hide bugs in post-inst scripts of the livepatch package.